pkg/kine/prepared/db.go

```diff
@@ -0,0 +1,233 @@
+}
+
+func (db *DB) Close() error {
+	db.mu.Lock()
+	defer db.mu.Unlock()
+
+	errs := []error{}
+	for len(db.store) != 0 {
+		if err := db.drop(db.lruList.tail()); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	if err := db.underlying.Close(); err != nil {
+		errs = append(errs, err)
+	}
+
+	return errors.Join(errs...)
+}
+
+func (db *DB) BeginTx(ctx context.Context, opts *sql.TxOptions) (*Tx, error) {
+	tx, err := db.underlying.BeginTx(ctx, opts)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Tx{
+		db: db,
+		tx: tx,
+	}, nil
+}
+
+func (db *DB) touch(entry *lruEntry) {
+	db.lruList.remove(entry)
+	db.lruList.add(entry)
+}
+
+func (db *DB) prepare(ctx context.Context, query string) (*stmt, error) {
+	if stmt := db.get(query); stmt != nil {
+		return stmt, nil
+	}
+	prepared, err := db.underlying.PrepareContext(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+	result := &stmt{prepared, atomic.Int32{}}
+	result.Ref()
+
+	db.put(query, result)
+	return result, nil
+}
+
+func (db *DB) get(key string) *stmt {
+	db.mu.Lock()
+	defer db.mu.Unlock()
+
+	if entry, ok := db.store[key]; ok {
+		db.touch(entry)
+		return entry.stmt.Ref()
+	}
+	return nil
+}
+
+func (db *DB) put(query string, stmt *stmt) {
+	if db.maxSize <= 0 {
+		return
+	}
+
+	db.mu.Lock()
+	defer db.mu.Unlock()
+
+	entry, ok := db.store[query]
+	if ok {
+		db.touch(entry)
+		entry.stmt.Close()
+		entry.stmt = stmt.Ref()
+		return
+	}
+
+	if len(db.store) >= db.maxSize {
+		db.drop(db.lruList.tail())
+	}
+
+	entry = &lruEntry{
+		query: query,
+		stmt:  stmt.Ref(),
+	}
+	db.store[query] = entry
+	db.lruList.add(entry)
+}
+
+func (db *DB) drop(entry *lruEntry) error {
+	db.lruList.remove(entry)
+	delete(db.store, entry.query)
+	return entry.stmt.Close()
+}
+
+type stmt struct {
+	*sql.Stmt
+	count atomic.Int32
+}
+
+func (s *stmt) Ref() *stmt {
+	s.count.Add(1)
+	return s
+}
+
+func (s *stmt) Close() error {
+	if s.count.Add(-1) <= 0 {
+		return s.Stmt.Close()
+	}
+	return nil
+}
+
+type Tx struct {
+	db *DB
+	tx *sql.Tx
+}
+
+func (tx *Tx) ExecContext(ctx context.Context, query string, args ...any) (sql.Result, error) {
+	stmt, err := tx.db.prepare(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+	defer stmt.Close()
+
+	return tx.tx.StmtContext(ctx, stmt.Stmt).ExecContext(ctx, args...)
+}
+
+func (tx *Tx) QueryContext(ctx context.Context, query string, args ...any) (*sql.Rows, error) {
+	stmt, err := tx.db.prepare(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+	defer stmt.Close()
+
+	return tx.tx.StmtContext(ctx, stmt.Stmt).QueryContext(ctx, args...)
+}
+
+func (tx *Tx) Commit() error   { return tx.tx.Commit() }
+func (tx *Tx) Rollback() error { return tx.tx.Rollback() }
+
+type lruList struct {
+	// lruList is implemented as a circular linked list
+	head *lruEntry
+}
+
+type lruEntry struct {
+	query string
+	stmt  *stmt
+
+	// Adjacent entries, unless at the tail of the list, the
+	// next entry was added less recently tham the current.
```

Comment:
```suggestion
	// next entry was added less recently than the current.
```

Reply:
All the lru has been removed as probably not very useful and yet quite complex.